services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartS3AsyncClient.java

```diff
@@ -46,32 +46,21 @@ public final class MultipartS3AsyncClient extends DelegatingS3AsyncClient {
 
     private static final ApiName USER_AGENT_API_NAME = ApiName.builder().name("hll").version("s3Multipart").build();
 
-    private static final long DEFAULT_MIN_PART_SIZE = 8L * 1024 * 1024;
-    private static final long DEFAULT_THRESHOLD = 8L * 1024 * 1024;
-    private static final long DEFAULT_API_CALL_BUFFER_SIZE = DEFAULT_MIN_PART_SIZE * 4;
-
     private final UploadObjectHelper mpuHelper;
     private final CopyObjectHelper copyObjectHelper;
 
     private MultipartS3AsyncClient(S3AsyncClient delegate, MultipartConfiguration multipartConfiguration) {
         super(delegate);
         MultipartConfiguration validConfiguration = Validate.getOrDefault(multipartConfiguration,
                                                                           MultipartConfiguration.builder()::build);
-        long minPartSizeInBytes = Validate.getOrDefault(validConfiguration.minimumPartSizeInBytes(),
-                                                        () -> DEFAULT_MIN_PART_SIZE);
-        long threshold = Validate.getOrDefault(validConfiguration.thresholdInBytes(),
-                                               () -> DEFAULT_THRESHOLD);
-        long apiCallBufferSizeInBytes = Validate.getOrDefault(validConfiguration.apiCallBufferSizeInBytes(),
-                                                              () -> computeApiCallBufferSize(validConfiguration));
+        MultipartConfigurationResolver resolver = new MultipartConfigurationResolver(validConfiguration);
+        long minPartSizeInBytes = resolver.minimalPartSizeInBytes();
+        long threshold = resolver.thresholdInBytes();
+        long apiCallBufferSizeInBytes = resolver.apiCallBufferSize();
```

Comment:
Should we just modify the MultipartConfiguration in-place when the client is created? I'm worried about a future bug where someone ends up using the MultipartConfiguration instead of the MultipartConfigurationResolver.

Reply:
The issue with that is if we are going to introduce a new way to configure settings, say, environment variables, we'd probably still need another layer to resolve it. This may also make it harder to integrate features like smart defaults because the default values depend on other settings that are not available in `MultipartConfiguration`. 
`MultipartConfiguration` returns boxed type, so hopefully that'd force the implementer to figure out they should use this class?

Reply:
I'm not so sure, but it seems like you've thought about it more than me. I'm fine with it, then.